Adding a feature to add column filters to Pinot query handlers.

buchi-busireddy opened the pull request:
This feature will be useful when there are two different Pinot views which serve the same set of attributes but they have different sets of data.
As an example, you could have all spans in one table but you can also have just entry spans in a different Pinot table, which is optimized for entry span queries. Without this feature, there is no way to select the correct view for the span queries in this example but with this feature, you can select the optimized view for entry span queries.

tim-mwangi commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/Params.java:
```diff
@@ -56,6 +56,58 @@ private Params(
+
+  @Override
+  public int hashCode() {
+    int result = integerParams.hashCode();
```
Why are you not using Objects methods?

buchi-busireddy replied:
Any advantages that gives specifically? I can check that.

tim-mwangi commented on query-service-impl/src/test/resources/log4j2.properties:
```diff
@@ -6,3 +6,6 @@ appender.console.layout.type=PatternLayout
 rootLogger.level=INFO
 rootLogger.appenderRef.stdout.ref=STDOUT
+
+logger.hypertrace.level = DEBUG
```
rollback to info

tim-mwangi replied:
oh never mind. this is impl.

tim-mwangi commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/Params.java:
```diff
@@ -56,6 +56,58 @@ private Params(
   }
 
+  @Override
+  public String toString() {
```
Should be easy to unit test these, no?

buchi-busireddy replied:
It's only used for logging, hence didn't bother to add the unit test. Equals is unit tested though.

tim-mwangi commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -147,6 +185,130 @@ public QueryCost canHandle(
+   * Method to check if the given ViewColumnFilter matches the given query filter. A match here
+   * means the view column is superset of what the query filter is looking for, not an exact match.
+   */
+  private boolean doesFiltersMatch(ViewColumnFilter viewColumnFilter, Filter queryFilter) {
```
doFiltersMatch?

buchi-busireddy replied:
Will do.

buchi-busireddy replied:
Done.

tim-mwangi commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -189,6 +358,46 @@ public void handleRequest(
+        return Filter.newBuilder(filter).clearChildFilter().addAllChildFilter(newFilters).build();
+      }
+    } else {
+      if (filter.getLhs().getColumnIdentifier().getColumnName().equals(columnName)) {
```
What if columnIdentifier is not set? Can that even happen with proto?

buchi-busireddy replied:
ideally not but I can add a requirement here so that we don't run into NPE.

aaron-steinfeld replied:
It can't - protos are never null!
```java
  public org.hypertrace.core.query.service.api.ColumnIdentifier getColumnIdentifier() {
    if (valueCase_ == 1) {
       return (org.hypertrace.core.query.service.api.ColumnIdentifier) value_;
    }
    return org.hypertrace.core.query.service.api.ColumnIdentifier.getDefaultInstance();
  }
```

buchi-busireddy replied:
Done.

tim-mwangi commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -147,6 +185,130 @@ public QueryCost canHandle(
+   * given query.
+   */
+  @Nonnull
+  private Map<String, Filter> getQueryFilterMap(QueryRequest request) {
```
You are not caring if it's an OR or AND filter?

buchi-busireddy replied:
I've thought about it. If we handle `OR` and `AND`, the behavior wouldn't change for `AND` but we can actually optimize for `OR` scenario. However, I wanted to avoid over-engineering and keep the logic simple for now but I could be convinced to optimize that though :)

aaron-steinfeld replied:
Haven't dove into this enough to know if I'm misunderstanding the comment, but OR vs AND seems like it should matter.

A view filtered to `FOO = x` should match a query with `FOO = x AND BAR = y` but can't match `FOO = X OR BAR = y` (since there could be spans with `BAR = Y` not present in this view). Is that distinction accounted for somewhere?

buchi-busireddy replied:
@aaron-steinfeld Good point, and that case is not handled with the current change. I was originally debating whether that validation is responsibility of the above layers or the view handler. Handling it here would let the view filter attribute to be treated like any other attribute so I like what you're suggesting.

It sounds like we have to check the entire query tree to make sure there is no `OR` in the hierarchy from the matching view filter node until the root. Is that what you're thinking too?

buchi-busireddy replied:
actually, nm about my second sentence. I got the case and I'm working on a fix for that. Thank you!

aaron-steinfeld replied:
I believe so, although difficult to put into words. More abstractly, any filter tree must be convertible to `<subsetOfviewFilter> && <restOfFilterTree>`.
That could happen in a two different ways, it's a fun tree problem. Recursing from root, a node matches if:
- It's AND and at least one child matches
- It's OR and every child matches

That way:
- Match: `(FOO = x AND BAR = y) OR (FOO = x AND BAR = z)
- No Match: `(FOO = x OR BAR = y) AND (FOO = x OR BAZ = z)

Just for fun... I went down a logic rabbit hole when writing examples, so I guess the above doesn't always hold. Take
`(FOO = x OR BAR = y) AND (FOO = x OR BAR = z)
  ```
(FOO = x OR BAR = y) AND (FOO = x OR BAR = z) -> (un)distribute
(FOO = x ) OR (BAR = y AND BAR = z) -> (mutually exclusive)
(FOO = x ) OR FALSE -> (identity)
FOO = x
  ```
 Not sure how deep we want to go with our boolean logic...

aaron-steinfeld replied:
Hah missed your second comment because I got carried away, but hopefully my comment is useful, else it was fun to write.

buchi-busireddy replied:
Implemented this and handling most (all?) of the cases that we discussed.

aaron-steinfeld commented:
Also just realized this will have some implications on attribute projections, as projections will need to be resolved before we do view filtering. I guess I'll do a full review.

aaron-steinfeld commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -155,6 +317,13 @@ public void handleRequest(
     long start = System.currentTimeMillis();
     validateQueryRequest(queryContext, request);
+
+    // Remove the filters which match the view column filters because they shouldn't be passed down
```
That's not strictly true. Take the case of a view filtered to `FOO IN (X, Y)` and a filter provided that matches the view, `FOO = X` - it needs to be rewritten to `FOO = Y` (or `FOO IN (Y)`) when calling pinot. We should either handle that case, or remove the subset match logic and only take exact matches.

buchi-busireddy replied:
Fixed this and added a test case to cover this case.

buchi-busireddy commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/ViewColumnFilter.java:
```diff
@@ -0,0 +1,39 @@
+
+  public static ViewColumnFilter from(Config filterConfig) {
+    Operator operator = Operator.valueOf(filterConfig.getString("operator"));
+    return new ViewColumnFilter(operator, new HashSet<>(filterConfig.getStringList("values")));
```
Yeah I was debating if I should use `value` for `EQ` and `values` for `IN` but I thought that might make it complex. I can fix it if you think that's fine, I don't have a strong feeling.

aaron-steinfeld replied:
I'd probably support either key in the config, reading as string for the single and list for the plural. You could then put either into the set and internally treat them as the same - that way we don't sacrifice the API for implementation complexity.

buchi-busireddy replied:
Addressed.

buchi-busireddy commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/ViewDefinition.java:
```diff
@@ -70,7 +74,15 @@ public static ViewDefinition parse(Config config, String tenantColumnName) {
+
+    final Map<String, ViewColumnFilter> filterMap = new HashMap<>();
+    if (config.hasPath("filters")) {
+      for (Config filterConfig: config.getConfigList("filters")) {
```
I'll add comments.

buchi-busireddy replied:
Done.

buchi-busireddy commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -147,6 +185,130 @@ public QueryCost canHandle(
+      return isSubSet(viewColumnFilter.getValues(), queryFilter.getRhs());
+    } else {
+      // query filter must be an equals filter in this case.
+      if (queryFilter.getOperator() != Operator.EQ) {
```
Done.

buchi-busireddy commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -147,6 +185,130 @@ public QueryCost canHandle(
+    while (!filters.isEmpty()) {
+      Filter filter = filters.poll();
+
+      // Check if the filter is a leaf filter or not.
```
Agree. I also realized a bit late. I think we can fix this, though in a different PR. Definitely fixing this will make it cleaner.

aaron-steinfeld replied:
yep, certainly separate from this PR.

buchi-busireddy replied:
Logged https://traceableai.atlassian.net/browse/ENG-6061 to take care of this

buchi-busireddy commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -147,6 +185,130 @@ public QueryCost canHandle(
+      if (filter.getChildFilterCount() > 0) {
+        filters.addAll(filter.getChildFilterList());
+      } else {
+        filterMap.put(filter.getLhs().getColumnIdentifier().getColumnName(), filter);
```
Fixed this now.

aaron-steinfeld commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -147,63 +148,110 @@ public QueryCost canHandle(
+        .collect(Collectors
+            .toMap(f -> f.getLhs().getColumnIdentifier().getColumnName(), f -> f));
+    if (childLeafFilterMap.isEmpty()) {
+      long matchCount = filter.getChildFilterList().stream()
```
If we flip the operator check, we can make this clearer by not using a match count but instead explicit predicates:

```
if (filter.getOperator() == Operator.AND) {
  return filter.getChildFilterList().stream().anyMatch(f -> doViewFiltersMatch(f, viewFilterMap));
} else if (filterOperator() === Operator.OR) {
  return filter.getChildFilterList().stream().allMatch(f -> doViewFiltersMatch(f, viewFilterMap));
}
```
(alternatively, drop the map above and use that in

aaron-steinfeld commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -147,63 +148,110 @@ public QueryCost canHandle(
+      return filter.getOperator() == Operator.OR && matchCount == filter.getChildFilterCount();
+    }
+
+    if (filter.getOperator() != Operator.AND) return false;
```
missing brackets. Also is this the case now where we have at least one leaf child? Why can't that match in the OR case? Admittedly, it's probably less common.
```
foo = x OR (foo = x AND bar = y) // Matches foo = x
```

aaron-steinfeld commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -147,63 +148,110 @@ public QueryCost canHandle(
+
+    if (filter.getOperator() != Operator.AND) return false;
+
+    for (Map.Entry<String, ViewColumnFilter> entry : viewFilterMap.entrySet()) {
```
We seem to go back and forth between for and stream quite a bit. This also appears to error if we don't have a child filter that references one of the view filters (as 225 would return null), which I would expect to be a regular false condition. I think we avoid it with one of the short circuits, but it doesn't seem necessary as it just makes things more brittle.
As a stream:
```
return viewFilterMap.entrySet().stream()
   .allMatch(viewFilterEntry -> 
childLeafFilterMap.containsKey(entry.getKey()) && doesSingleViewFilterMatchLeafQueryFilter(entry.getValue(), childLeafFilterMap.get(entry.getKey())));
```

buchi-busireddy commented on query-service-impl/src/test/java/org/hypertrace/core/query/service/QueryRequestBuilderUtils.java:
```diff
@@ -38,18 +39,31 @@ public static Filter createEqualsFilter(String column, String value) {
   public static Expression createStringLiteralValueExpression(String value) {
     return Expression.newBuilder().setLiteral(LiteralConstant.newBuilder().setValue(
-        Value.newBuilder().setString(value))).build();
+        Value.newBuilder().setString(value).setValueType(ValueType.STRING))).build();
```
This was written long time ago and I don't remember if there was any reason. It's just that we got started there and continued that way AFAIK.

aaron-steinfeld replied:
Looking at value and trying to reach way back into the mists of the past, I believe it was because oneof doesn't support repeated and maybe map fields. A wrapper would also have sufficed there, but that was likely the reason.

buchi-busireddy commented:
@tim-mwangi @aaron-steinfeld Addressed all the comments or otherwise replied to them. Added unit tests for different scenarios. This is ready for review again.

aaron-steinfeld commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -171,61 +168,42 @@ public QueryCost canHandle(QueryRequest request, Set<String> referencedSources,
-          // Few columns matched but others didn't match.
-          return false;
-        }
+        return filter.getOperator() == Operator.AND && matchCount == viewFilterMap.size();
```
This is checking for a "full match" but misses a number of scenarios. All we know about these children is that at least one is a leaf, but not necessarily all.
In the AND case, we need each view filter to be matched, but that could happen in one or more nodes. It's also not necessarily indicated by the count as we could have duplicates:

Duplicate case:
View Filter: X = 'A' AND Y = 'B'
Query Filter: X = 'A' AND X= 'A'

Single match case (this specific example wouldn't happen given our optimize logic, but are we intending on relying on that here? that seems like a far distance for an assumption to travel):
View Filter: X = 'A' AND Y = 'B'
Query Filter: (X = 'A' AND Y= 'B') AND Z = 'C'

In the OR case, we need each child to be a full match, but right now we'd never match since we only look at ANDs

Each full match case:
View Filter: X = 'A'
Query Filter: X = 'A' OR X = 'A' // many equivalencies to this

buchi-busireddy replied:
Yes, I'm relying on the filter optimization for all these 3 cases and other similar duplicate cases or flattening predicates. I feel that keeps the tree navigation logic much simpler instead of handling so many cases.
Reg keeping the filter optimization far from this code: I thought about it and the reason I'm doing it early on is 
1. so that we don't have to repeat it in each handler
2. other parts of the code like conversion to SQL can take advantage of it too.

The optimization logic is in a separate class and can be invoked from here too if needed but I think this is fine.

aaron-steinfeld replied:
See the later part of my most recent comment - https://github.com/hypertrace/query-service/pull/17#discussion_r489412091  - I believe this proposes a way of doing this in a much more straightforward way that doesn't require (but works with) optimization and all scenarios.

aaron-steinfeld commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/ViewColumnFilter.java:
```diff
@@ -26,7 +39,11 @@ public ViewColumnFilter(Operator operator, Set<String> values) {
+    if (operator == Operator.EQ) {
+      return new ViewColumnFilter(operator, Set.of(filterConfig.getString("value")));
+    } else {
+      return new ViewColumnFilter(operator, new HashSet<>(filterConfig.getStringList("values")));
```
`Set.copyOf` to match 43, and keep things immutable

buchi-busireddy replied:
Done.

buchi-busireddy commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -171,61 +168,42 @@ public QueryCost canHandle(QueryRequest request, Set<String> referencedSources,
-    // Navigate the entire filter tree to check if every query filter which has the
-    // view filter columns has all matching view filters. If not, return false since
-    // this view can't handle such query.
+    // 1. Basic case: Filter is a leaf node. This will be a match only if there is only one
```
Case 3 handles the example you have given here. In last iteration I was checking for viewFilter count but I found this logic to be simpler and cleaner.

The approach doesn't support partial match. Are there any pieces of code which seem to indicate that? I can add comments/clarify if needed but the whole logic only checks for full matches. Updating the javadoc of different classes and methods to make this clear.

Reg the other approach you're suggesting, the v1 code I've written was on those same lines but what complicated the solution is, there could be multiple matching filter nodes in the tree because of duplicate predicates. So, we should be able to return multiple matching filter nodes. I can take another fresh look at that approach in another PR but I think this can go because it's functionally correct and code is easy enough.

aaron-steinfeld replied:
> Case 3 handles the example you have given here. In last iteration I was checking for viewFilter count but I found this logic to be simpler and cleaner.

Maybe I'm misreading this, but given a single leaf filter regardless of how many view filters we have, we'd return true on 192 and go into the singleViewMatch function. That function would only check if the leaf filter matches the corresponding col view filter - in other words, a partial match. Later, we rely on this returning true for a partial match, when we do things like get the match count and make sure it matches the number of view cols (line 212).

> Reg the other approach you're suggesting, the v1 code I've written was on those same lines but what complicated the solution is, there could be multiple matching filter nodes in the tree because of duplicate predicates. So, we should be able to return multiple matching filter nodes. I can take another fresh look at that approach in another PR but I think this can go because it's functionally correct and code is easy enough.

Not sure I follow the issue here - if we're maintaining a set of matches, duplicates wouldn't matter. I don't believe this is functionally correct, let's talk it through.

buchi-busireddy commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -171,61 +168,42 @@ public QueryCost canHandle(QueryRequest request, Set<String> referencedSources,
+      } else {
+        // 3. An internal filter with children which are leaves: All the view filters
+        // should be matched and AND'ed, otherwise it's not a match.
+        long matchCount = filter.getChildFilterList().stream()
```
Fixed.

buchi-busireddy commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -171,61 +168,42 @@ public QueryCost canHandle(QueryRequest request, Set<String> referencedSources,
       if (childLeafFilterMap.isEmpty()) {
         long matchCount = filter.getChildFilterList().stream()
             .map(f -> doViewFiltersMatch(f, viewFilterMap))
             .filter(b -> b).count();
```
Fixed.

aaron-steinfeld commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -189,6 +369,58 @@ public void handleRequest(
+      Map<String, ViewColumnFilter> columnFilterMap) {
+
+    Filter newFilter = removeViewColumnFilter(request.getFilter(), columnFilterMap);
+    QueryRequest.Builder builder = QueryRequest.newBuilder(request).clearFilter();
+    if (!Filter.getDefaultInstance().equals(newFilter)) {
+      builder.setFilter(newFilter);
+    }
+    return builder.build();
```
Query should work just fine with a default empty filter (as that's what getFilter will always return anyway, never null)
```suggestion
    return QueryRequest.newBuilder(request).setFilter(newFilter).build();
```

aaron-steinfeld commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -189,6 +369,58 @@ public void handleRequest(
+      Set<Filter> newFilters = filter.getChildFilterList().stream()
+          .map(f -> removeViewColumnFilter(f, columnFilterMap))
+          .filter(f -> !Filter.getDefaultInstance().equals(f))
+          .collect(Collectors.toCollection(LinkedHashSet::new));
```
nit: if we don't have a particular need for a linked hash set, there's shortcuts like `Collectors.toSet()` or `Collectors.toUnmodifiableSet()`

buchi-busireddy replied:
I wanted to retain the order, mainly because I was verifying sql query in the unit test and without order the predicates were sometimes coming up in different order.

aaron-steinfeld commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -189,6 +369,58 @@ public void handleRequest(
+      if (newFilters.isEmpty()) {
+        return Filter.getDefaultInstance();
+      } else if (newFilters.size() == 1) {
+        return newFilters.stream().findFirst().get();
```
nit - believe we're already using guava in here, so a little clearer to do:
```suggestion
        return Iterables.getOnlyElement(arguments);
```

buchi-busireddy replied:
Oh this is cool. Will fix it.

aaron-steinfeld commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -189,6 +369,58 @@ public void handleRequest(
+      return queryFilter;
+    } else {
+
+      // The only case where we need to return non-null filter is when view filter is based on 'IN'
```
are there cases here with different RHS? like:
VF: `X = 'A'`
QF: `X = 'A' AND X ='B'` - should return nothing

VF: `X IN ('A', 'B')`
QF: `X IN ('A', 'B') AND X ='B'` - should return only B matches

aaron-steinfeld replied:
@buchi-busireddy - this is the last comment I think we need to make sure we resolve (either through comment or fix)

buchi-busireddy replied:
You need to get 5 ⭐ for your reviews 👍 

I've fixed this case and added unit test also now.

aaron-steinfeld commented on query-service-impl/src/main/java/org/hypertrace/core/query/service/pinot/PinotBasedRequestHandler.java:
```diff
@@ -186,31 +177,26 @@ public QueryCost canHandle(QueryRequest request, Set<String> referencedSources,
-  private boolean doViewFiltersMatch(Filter filter, Map<String, ViewColumnFilter> viewFilterMap) {
-    // 1. Basic case: Filter is a leaf node. This will be a match only if there is only one
-    // view filter and that matches the leaf filter.
+  private Set<String> getMatchingViewFilterColumns(Filter filter, Map<String,
```
beautiful!

buchi-busireddy replied:
Thanks for the thorough review and ideas.

aaron-steinfeld reviewed (changes requested):
Got excited and approved with the rewrite, but there's one important unresolved comment so removing (can't figure out how to just remove approval)

tim-mwangi reviewed (approved):
I'm gonna assume this was rigorously reviewed 😂
